### DISCUSSION TITLE
update src attribute using event

### DISCUSSION THREAD

**Discussion by ghost** [thephpleague/commonmark#647] [Q&A] [answered]
Using the commonmark package the first time this excessive.

I've used the code to [add a class to an image](https://github.com/GrahamCampbell/Laravel-Markdown/issues/58) to make images hidden, since we are facing much abuses when users are allowed to inlclude images by markdown.

The idea was: 

- register event listener, hook up into the `DocumentParsedEvent`
- modify the `src` element of every image instance (I've taken the idea from the render part [here](https://github.com/thephpleague/commonmark/blob/master/src/Inline/Renderer/ImageRenderer.php#L48))

ImageManipulatorProcessor

```
public function __invoke(DocumentParsedEvent $e)
    {
        $walker = $e->getDocument()->walker();

        while ($event = $walker->next()) {
            $node = $event->getNode();

            if (!($node instanceof Image) || !$event->isEntering()) {
                continue;
            }

            if ($node instanceof Image) {
                $node->data['attributes']['src'] = '';
            }
        }
    }
```

Is it not possible to change already set attributes after they are rendered? How to update the `src` attribute if so?

**Comment by colinodell** [comment, marked as answer]
The `src` is not stored in the `attributes` array - `Image` extends from [AbstractWebResource](https://github.com/thephpleague/commonmark/blob/master/src/Inline/Element/AbstractWebResource.php) which stores the URL in its own field.  Your code should work with a simple change:

```diff
public function __invoke(DocumentParsedEvent $e)
    {
        $walker = $e->getDocument()->walker();

        while ($event = $walker->next()) {
            $node = $event->getNode();

            if (!($node instanceof Image) || !$event->isEntering()) {
                continue;
            }

            if ($node instanceof Image) {
-               $node->data['attributes']['src'] = '';
+               $node->setUrl('');
            }
        }
    }
```

Alternatively, if your goal is to completely remove the image, you could try removing the whole image element from the AST here (though that might be complex to implement since I don't think `$node->detach()` plays nicely with the walker).

Another option would be replacing the image renderer with something that always returns an empty string.

Or you could disable/remove the `BangParser` so that `!` isn't treated as a special character.

I hope this helps!

**Comment by ghost** [comment]
@colinodell Thanks, working fine, I'll stay with the replacing function of the `src` attribute. I think this will be the easiest way at the moment, so I can stay using the `DocumentParsedEvent`.
